### PR TITLE
update referrer manifest descriptor size

### PR DESCRIPTION
**Pull request by MinerYang**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
https://github.com/goharbor/harbor/issues/20187
```
curl -k -u 'admin:pwd' https://xx.xxx.xxx.xx/v2/library/hello-world/referrers/sha256:d37ada95d47ad12224c205a938129df7a3e52345828b4fa27b03a98825d1e2e7 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   300  100   300    0     0  26073      0 --:--:-- --:--:-- --:--:-- 27272
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:8a12406f6f62e297a93400610fa6a136c7fcf6a0956219bc929676014d2f2fc3",
      "size": 1620,
      "artifactType": "application/vnd.oci.image.config.v1+json"
    }
  ]
}
```
```
/data/registry/docker/registry/v2/blobs/sha256$ wc -c <  8a/8a12406f6f62e297a93400610fa6a136c7fcf6a0956219bc929676014d2f2fc3/data
1620
```

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/server/registry/referrers.go [hunk 1]
```diff
@@ -84,9 +85,16 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 			lib_http.SendError(w, err)
 			return
 		}
+		// pull manifest
+		mani, _, err := reg.Cli.PullManifest(accArt.RepositoryName, acc.GetData().Digest)
```
Comment (wy65701436) [inline src/server/registry/referrers.go:89]: Please try to fetch from redis first, if no luck, then use the pulling.

File: src/server/registry/referrers.go [hunk 2]
```diff
@@ -75,29 +82,51 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		lib_http.SendError(w, err)
 		return
 	}
-
+	fmt.Printf("accs length is %d\n", len(accs))
```
Comment (stonezdj) [inline src/server/registry/referrers.go:85]: please remove the print, use log.Debug if required

File: src/server/registry/referrers.go [hunk 3]
```diff
@@ -75,29 +84,65 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+				}
+			}
+		}
+		if !fromCache {
+			mani, _, err := r.registryClient.PullManifest(accArt.RepositoryName, accArtDigest)
+			if err != nil {
+				lib_http.SendError(w, err)
+				return
+			}
+			_, maniContent, _ = mani.Payload()
```
Comment (wy65701436) [inline src/server/registry/referrers.go:121]: error handling?

File: src/server/registry/referrers.go [hunk 4]
```diff
@@ -75,29 +84,69 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+
+		// pull manifest, will try to pull from cache first
+		// and write to cache when pull manifest from registry at first time
+		if config.CacheEnabled() {
+			maniContent, err = r.maniCacheManager.Get(req.Context(), accArtDigest)
+			if err == nil {
+				fromCache = true
+			} else {
+				log.Debugf("failed to get manifest %s from cache, will fallback to registry, error: %v", accArtDigest, err)
+				if strings.Contains(err.Error(), "not found") {
```
Comment (wy65701436) [inline src/server/registry/referrers.go:110]: errors.As(cache.ErrNotFound)

File: src/server/registry/referrers.go [hunk 5]
```diff
@@ -75,29 +84,69 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-		accArt, err := r.artifactManager.GetByDigest(ctx, repository, acc.GetData().Digest)
+		accArtDigest := acc.GetData().Digest
+		accArt, err := r.artifactManager.GetByDigest(ctx, repository, accArtDigest)
 		if err != nil {
 			lib_http.SendError(w, err)
 			return
 		}
-		mf := ocispec.Descriptor{
+		// whether get manifest from cache
+		fromCache := false
```
Comment (wy65701436) [inline src/server/registry/referrers.go:97]: the fromCache confuses me, the exactly meaning of this variable here is to success read manifest from cache.  It is not pointing whether to get manifest from cache

File: src/server/registry/referrers.go [hunk 6]
```diff
@@ -75,29 +84,69 @@ func (r *referrersHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+				lib_http.SendError(w, err)
+				return
+			}
+			// write manifest to cache when first time pulling
+			if writeCache {
+				err = r.maniCacheManager.Save(req.Context(), accArtDigest, maniContent)
+				if err != nil {
+					log.Warningf("failed to save accArt manifest %s to cache, error: %v", accArtDigest, err)
+				} else {
+					log.Debugf("write accArt manifest to cache successfully: %s", accArtDigest)
```
Comment (wy65701436) [inline src/server/registry/referrers.go:132]: no need to log this.

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/server/registry/referrers.go (hunk 5 above)
Comment (MinerYang) [reply src/server/registry/referrers.go:97]: When user not enable cache layer config  or read manifest from cache failed, it need pull manifest which would be `not fromCache`.
Only when Cache enabled and successfully read manifest from redis, fromCache would be true.

**Review by stonezdj — APPROVED** [review]

**Pull request merged**